### PR TITLE
Bookmarked portals

### LINKED ISSUE #27: Allow export from bookmarked portals
To be able to create some multilayered BAFs 😸

### PR DESCRIPTION
**Pull request by pmodin**
closes #27 

- [x] Works to export directly to IMF
- [x] Works without any bookmarked portals
- [x] Works without bookmark plugin (just display a hint that the feature is available)

### CODE REVIEW

File: IngressMaxFields.user.js [hunk 1]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+        for (var folder in folders) {
+            data += `<div onClick='window.plugin.ingressmaxfield.appendBookmarkFolder("${folder}");'>${folders[folder]}</div>`
+        }
+
+        return data;
+    }
+
+    // Generate string for given bookmarked portal
+    self.genStrFromBookmarkPortal = function genStrFromBookmarkPortal(portal, folder) {
+        var portalsList = JSON.parse(localStorage['plugin-bookmarks']);
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l149)

File: IngressMaxFields.user.js [hunk 2]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+        var portal = portalsList.portals[folder]["bkmrk"][portal];
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: ["portal" is already defined](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l150)

File: IngressMaxFields.user.js [hunk 3]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+    // Return a list of portal bookmark folders
+    self.checkBookmarks = function checkBookmarks() {
+        if (!window.plugin.bookmarks) {
+            return null;
+        }
+
+        var portalsList = JSON.parse(localStorage['plugin-bookmarks']);
+
+        var res = {};
+        for (var folder in portalsList.portals) {
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l275)

File: IngressMaxFields.user.js [hunk 4]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+            data += self.genStrFromBookmarkPortal(portal, folder) + "\n";
+        }
+
+        return data;
+    }
+
+
+    // Write a list of all portal bookmarks to the text area
+    self.appendBookmarkFolder = function appendBookmarkFolder(folder) {
+        var $form = $('form[name="maxfield"] textarea')
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l171)

File: IngressMaxFields.user.js [hunk 5]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+        var portal = portalsList.portals[folder]["bkmrk"][portal];
+        var latlng = portal.latlng.split(',');
+
+        return self.genStr(portal.label, latlng[0], latlng[1], portal.guid);
+    }
+
+    // Generate string for all the portals
+    self.genStrFromBookmarkFolder = function genStrFromBookmarkFolder(folder) {
+        var data = "",
+            portalsList = JSON.parse(localStorage['plugin-bookmarks']);
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l159)

File: IngressMaxFields.user.js [hunk 6]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
-	data += "    </table>";
-	data += "</div>";
-	data += "</div>";
-	data += "</form>";
+
+    // Return markup for the bookmarks to show in the IMF dialog
+    self.renderPortalBookmarkFolders = function renderPortalBookmarkFolder(folders) {
+        var data = "<p>With Portal Bookmarks, you can populate the list with a folder you have created instead:</p>";
+
+        for (var folder in folders) {
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l140)

File: IngressMaxFields.user.js [hunk 7]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+            data += `<div onClick='window.plugin.ingressmaxfield.appendBookmarkFolder("${folder}");'>${folders[folder]}</div>`
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l141)

File: IngressMaxFields.user.js (hunk 4 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l171)

File: IngressMaxFields.user.js [hunk 8]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+        }
+
+        return data;
+    }
+
+    // Generate string for given bookmarked portal
+    self.genStrFromBookmarkPortal = function genStrFromBookmarkPortal(portal, folder) {
+        var portalsList = JSON.parse(localStorage['plugin-bookmarks']);
+        var portal = portalsList.portals[folder]["bkmrk"][portal];
+        var latlng = portal.latlng.split(',');
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l151)

File: IngressMaxFields.user.js [hunk 9]
```diff
@@ -133,47 +132,152 @@ function wrapper() {
...
+
+        return self.genStr(portal.label, latlng[0], latlng[1], portal.guid);
+    }
+
+    // Generate string for all the portals
+    self.genStrFromBookmarkFolder = function genStrFromBookmarkFolder(folder) {
+        var data = "",
+            portalsList = JSON.parse(localStorage['plugin-bookmarks']);
+
+        for (var portal in portalsList.portals[folder]["bkmrk"]) {
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651000314/issues/source?bid=3039347&fileBranchId=3324533#l161)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- IngressMaxFields.user.js  5
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

File: IngressMaxFields.user.js (hunk 2 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: ["portal" is already defined](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l150)

File: IngressMaxFields.user.js (hunk 1 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l149)

File: IngressMaxFields.user.js (hunk 3 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l275)

File: IngressMaxFields.user.js (hunk 4 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l171)

File: IngressMaxFields.user.js (hunk 5 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l159)

File: IngressMaxFields.user.js (hunk 6 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l140)

File: IngressMaxFields.user.js (hunk 7 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l141)

File: IngressMaxFields.user.js (hunk 4 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l171)

File: IngressMaxFields.user.js (hunk 8 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l151)

File: IngressMaxFields.user.js (hunk 9 above)
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651002189/issues/source?bid=3039347&fileBranchId=3324533#l161)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- IngressMaxFields.user.js  5
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

File: IngressMaxFields.user.js [hunk 10]
```diff
@@ -96,23 +87,31 @@ function wrapper() {
...
+        return str;
     };
 
+    self.genStrFromPortal = function genStrFromPortal(portal, portalGuid) {
+        var lat = portal._latlng.lat,
+            lng = portal._latlng.lng,
+            title = portal.options.data.title || "untitled portal";
+
+        return self.genStr(title, lat, lng, portalGuid);
+    }
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651008646/issues/source?bid=3039347&fileBranchId=3324533#l110)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- IngressMaxFields.user.js  8
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

File: IngressMaxFields.user.js [hunk 11]
```diff
@@ -133,47 +132,158 @@ function wrapper() {
...
+        var data = "<p>With Portal Bookmarks, you can populate the list with a folder you have created instead:</p>";
+
+        for (var folder in folders) {
+            if (folders.hasOwnProperty(folder)) {
+                data += `<div onClick='window.plugin.ingressmaxfield.appendBookmarkFolder("${folder}");'>${folders[folder]}</div>`;
+            }
+        }
+
+        return data;
+    }
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651008842/issues/source?bid=3039347&fileBranchId=3324533#l147)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- IngressMaxFields.user.js  8
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

File: IngressMaxFields.user.js [hunk 12]
```diff
@@ -133,47 +132,158 @@ function wrapper() {
...
+        var portalsList = JSON.parse(localStorage["plugin-bookmarks"]);
+
+        var res = {};
+        for (var folder in portalsList.portals) {
+            if (portalsList.portals.hasOwnProperty(folder)) {
+                res[folder] = portalsList.portals[folder].label;
+            }
+        }
+        return res;
+    }
```
Comment (itayo) [inline IngressMaxFields.user.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/file/2651008873/issues/source?bid=3039347&fileBranchId=3324533#l285)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- IngressMaxFields.user.js  8
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- IngressMaxFields.user.js  8
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

**Comment by itayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- IngressMaxFields.user.js  8
         
```


See the complete overview on [Codacy](https://www.codacy.com/app/stefan-nygren/IITC-Ingress-Maxfields-Exporter/pullRequest?prid=245831&bid=3324533)

**Comment by pmodin** [comment]
Regarding codacy, I can't see how to add functionality and keep complexity down without introducing a build step 😄 

Regarding your June without computer, I've pushed to pmodin/IITC-Ingress-Maxfields-Exporter until you're back. Have a nice vacation!

**Comment by WheelOfTheLaw** [comment]
Well done, works flawlessly so far

**Pull request merged**